lib/ts/controllers/s-modal.ts

```diff
@@ -0,0 +1,222 @@
+namespace Stacks {
+
+    export class ModalController extends Stacks.StacksController {
+        static targets = ["modal"];
+
+        private modalTarget!: HTMLElement;
+
+        private _boundClickFn!: any;
+        private _boundKeypressFn!: any;
+
+        private returnElement!: HTMLElement;
+
+        private _boundTabTrap!: any;
+
+        connect () {
+            this._validate();
+        }
+
+        /**
+         * Disconnects all added event listeners on controller disconnect
+         */
+        disconnect() {
+            this._unbindDocumentEvents();
+        };
+
+        /**
+         * Toggles the visibility of the modal
+         */
+        toggle () {
+            this._toggle();
+        }
+
+        /**
+         * Shows the modal
+         */
+        show () {
+            this._toggle(true);
+        }
+
+        /**
+         * Hides the modal
+         */
+        hide () {
+            this._toggle(false);
+        }
+
+        /**
+         * Validates the modal settings and attempts to set necessary internal variables
+         */
+        _validate() {
+            // check for returnElement support
+            var returnElementSelector = this.data.get("return-element");
+            if (returnElementSelector) {
+                this.returnElement = <HTMLElement>document.querySelector(returnElementSelector);
+
+                if (!this.returnElement) {
+                    throw "Unable to find element by return-element selector: " + returnElementSelector;
+                }
+            }
+        }
+
+        /**
+         * Toggles the visibility of the modal element
+         * @param show Optional parameter that force shows/hides the element or toggles it if left undefined
+         */
+        _toggle (show?: boolean | undefined) {
```

Comment:
I'm keeping it as-is for right now. The `_toggle` function is currently fairly symmetric and I don't want to tear it apart since we're trying to merge + release. I did add in the code shown to not change the modal state if the current state matches.